### ISSUE TITLE
receive: Remove MultiTSDBStore and replace with ProxyStore

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2864]
Those two structs are quite complex, but are doing exactly the same thing. 

Few (smallish) work Items to make that happen:

* rename `storeRef` to `cachedStoreClient`
* create `localStoreClient` that will allow calling locally some `storepb.StoreServer`. For receive, it would be TSDBStores. This will need to use our `tenantSeriesSetServer` or something similar.
* ensure similar spans for tracing and error logging with tenant ids

## Motivation

Maintaining two literally same things is tedious, we can simplify things :hugs: Especially that we have nice timeout, chunk dedup etc logics for proxy store we might want to use.

## Action Items:

* Thanos Receive is using ProxyStore instead of MultiTSDBStore for fanout Series request to TSDBs\
* MultiTSDB benchmark exists, just is swapped with ProxyStore
* Benchmarks before & after is produced, following script can be used:

```
#!/usr/bin/env bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

cd ${DIR}/../

NUM=${1}
RUN="${NUM}-receiveseries"
mkdir -p ${DIR}/bench_outs/${RUN}

for TEST in BenchmarkMultiTSDBSeries
do
echo "Running ${TEST} run ${RUN}"
go test -bench=${TEST} -run=^$ -benchmem -memprofile ${DIR}/bench_outs/${RUN}/memprofile${TEST//\//-}.out -timeout 2h -benchtime 30s ./pkg/store | tee  ${DIR}/bench_outs/${RUN}/bench${TEST//\//-}.out
done

PREV=$((NUM - 1))
benchstat -delta-test none ${DIR}/bench_outs/${PREV}-receiveseries/BenchmarkMultiTSDBSeries.out ${DIR}/bench_outs/${RUN}/BenchmarkMultiTSDBSeries.out | tee ${DIR}/bench_outs/${RUN}/diff.out

```

cc @brancz @kakkoyun

**Comment by brancz** [comment]
There were a few things that I ran into that I felt made this forced and caused odd interface implementations to make use of a "generic" abstraction. If you can come up with something reasonable though I'm all for deduplicating.

**Comment by bwplotka** [comment]
Yea, I think you might mean `Client` interface. This is bit more wider than gRPC StoreAPI Server, but still easy to implement something for local TSDBStore to fit (:  

Help wanted!

**Comment by brancz** [comment]
Yeah I still think it needs some refactoring I wouldn't want to press it into an interface that doesn't actually fit, the `Addr()` function for example makes no sense here. Again, I'm not against it, I just think it needs a larger refactoring, then I think this could work nicely.

**Comment by bwplotka** [comment]
Still todo.

**Comment by bwplotka** [comment]
Still valid (:

Kind Regards,
Bartek Płotka (@bwplotka)

**Comment by Abhishek357** [comment]
still valid

**Comment by yeya24** [comment]
Still valid I think.
Something really nice to support this is to have tenant label matching and min/max label filtering at the `Series` grpc request. This could make the receiver query faster.

**Comment by matej-g** [comment]
I looked into this TODO as a part of working on something else, I think should be doable (and yes, some adjustments to the client interface would be nice to make this fit more reasonably). I'll try to come back with a draft PR.

**Comment by matej-g** [comment]
Still relevant (still not done due to lack of time :crying_cat_face:)
